// @middleware.io/sourcemap-uploader/src/lib.ts
import { basename, isAbsolute, join } from "path";
import { cwd } from "process";
import { readFileSync, statSync } from "fs";
import { globSync } from "glob";
import fetch from "cross-fetch";
import { unlink } from "fs/promises";

type sourceMapOpts = {
  apiKey: string;
  appVersion: string;
  path: string;
  basePath: string;
  backendUrl?: string;
  allowNoop?: boolean;
  deleteAfterUpload?: boolean;
};

async function getAllSourceMapFiles(
  paths: string[],
  { allowNoop }: { allowNoop?: boolean }
) {
  const map: { path: string; name: string }[] = [];

  await Promise.all(
    paths.map(async (path) => {
      const realPath = isAbsolute(path) ? path : join(cwd(), path);

      if (statSync(realPath).isFile()) {
        map.push({
          path: realPath,
          name: basename(realPath),
        });

        return;
      }

      if (
        !allowNoop &&
        !globSync("**/*.{js,css}.map", {
          cwd: realPath,
          nodir: true,
          ignore: "**/node_modules/**/*",
        }).length
      ) {
        throw new Error(
          "No .js.map or .css.map files found. Please double check that you have generated sourcemaps for your app."
        );
      }

      for (const file of globSync("**/*.{js,css}.map", {
        cwd: realPath,
        nodir: true,
        ignore: "**/node_modules/**/*",
      })) {
        map.push({
          path: join(realPath, file),
          name: file,
        });
        const routeGroupRemovedPath = file.replaceAll(
          new RegExp(/(\(.+?\))\//gm),
          ""
        );
        if (file !== routeGroupRemovedPath) {
          // also upload the file to a path without the route group for frontend errors
          map.push({
            path: join(realPath, file),
            name: routeGroupRemovedPath,
          });
        }
      }
    })
  );

  return map;
}

async function uploadFile(filePath: string, uploadUrl: string, name: string) {
  const fileContent = readFileSync(filePath);
  const res = await fetch(uploadUrl, {
    method: "put",
    headers: {
      "x-ms-blob-type": "BlockBlob",
    },
    body: new Uint8Array(fileContent),
  });
  console.log(`[MW] Uploaded ${filePath} to ${name}`);
}

const makeFetch = async (
  url: string,
  apiKey: string,
  type: "POST" | "PUT",
  body: any
) => {
  const response = await fetch(url, {
    method: type,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`,
    },
    body: Buffer.from(JSON.stringify(body)).toString("base64"),
  });

  return await response.json();
};

export const uploadSourcemaps = async ({
  apiKey,
  appVersion,
  path,
  basePath,
  backendUrl,
  allowNoop,
  deleteAfterUpload,
}: sourceMapOpts) => {
  try {
    if (!apiKey || apiKey === "")
      if (process.env.MW_API_KEY) apiKey = process.env.MW_API_KEY;
      else throw new Error("api key cannot be empty");

    const backend =
      backendUrl || "https://app.middleware.io/api/v1/rum/getSasUrl";

    console.info(`Starting to upload source maps from ${path}`);

    const fileList = await getAllSourceMapFiles([path], { allowNoop });

    if (fileList.length === 0)
      throw new Error(
        `Error: No source maps found in ${path}, is this the correct path?`
      );

    // fetch the pre-signed url
    const res = await makeFetch(backend, apiKey, "POST", {
      fileNames: fileList.map(({ name }) => getBucketKey(basePath || "", name)),
      version: appVersion ?? "latest",
    });

    if (!res?.status || !res?.data || res.data?.length === 0)
      throw new Error("Error: Unable to generate source map upload urls.");

    await Promise.all(
      fileList.map(({ path, name }, idx) =>
        uploadFile(path, res.data[idx], name)
      )
    );
    if(deleteAfterUpload) {
      await Promise.all(
        fileList.map(({ path, name }, idx) => {
          if (path.endsWith(".map")) {
            unlink(path);
          }
        })
      );
    }
  } catch (e) {
    console.error(e);
    console.info("Failed to upload source maps. Please see reason above.");
    return;
  }
};

function getBucketKey(basePath: string, fileName: string) {
  const normalizedFilename = fileName.replace(/[()]/g, "");
  return `${basePath}${normalizedFilename}`;
}
