// @middleware.io/sourcemap-uploader/src/vite-plugin.ts
import type { Plugin, ResolvedConfig } from "vite";
import { relative, resolve } from "path";
import { cwd } from "process";
import { uploadSourcemaps } from "./lib.js";

export interface MiddlewareVitePluginOptions {
  /** The Middleware RUM API key (or set MW_API_KEY env var) */
  apiKey: string;
  /** The current version of your deploy */
  appVersion: string;
  /** Directory containing sourcemaps (defaults to Vite's build.outDir) */
  path?: string;
  /** Optional base path for the sourcemaps in the bucket */
  basePath?: string;
  /** Optional backend URL for self-hosted deployments */
  backendUrl?: string;
  /** Delete sourcemaps after uploading (default: false) */
  deleteAfterUpload?: boolean;
}

export default function middlewareVitePlugin(
  options: MiddlewareVitePluginOptions
): Plugin {
  const {
    apiKey,
    appVersion,
    basePath = "",
    backendUrl,
    deleteAfterUpload = false,
  } = options;

  let resolvedConfig: ResolvedConfig;

  return {
    name: "middleware-sourcemap-upload",
    apply: "build",
    configResolved(config) {
      resolvedConfig = config;
    },
    async closeBundle() {
      try {
        const outDir = resolve(resolvedConfig.root, resolvedConfig.build.outDir);
        const path =
          options.path ?? (relative(cwd(), outDir) || ".");

        await uploadSourcemaps({
          apiKey,
          appVersion,
          path,
          basePath,
          backendUrl,
          allowNoop: true,
          deleteAfterUpload,
        });
      } catch (e) {
        console.error(`Middleware failed to upload sourcemaps: ${e}`);
      }
    },
  };
}
