// @middleware.io/sourcemap-uploader/src/webpack-plugin.ts
import { uploadSourcemaps } from "./lib.js";

export default class MiddlewareWebpackPlugin {
  apiKey: string;
  appVersion: string;
  path: string;
  basePath: string;
  backendUrl: string;
  deleteAfterUpload?: boolean;

  constructor(
    apiKey: string,
    appVersion: string,
    path: string,
    basePath: string,
    backendUrl: string,
    deleteAfterUpload?: boolean
  ) {
    this.apiKey = apiKey;
    this.appVersion = appVersion;
    this.path = path;
    this.basePath = basePath;
    this.backendUrl = backendUrl;
    this.deleteAfterUpload = deleteAfterUpload;
  }

  apply(compiler: any) {
    compiler.hooks.afterEmit.tap("MiddlewareWebpackPlugin", async () => {
      try {
        await uploadSourcemaps({
          apiKey: this.apiKey,
          appVersion: this.appVersion,
          path: this.path,
          basePath: this.basePath,
          backendUrl: this.backendUrl,
          allowNoop: true,
          deleteAfterUpload: this.deleteAfterUpload,
        });
      } catch (e) {
        console.error(`Middleware failed to upload sourcemaps: ${e}`);
      }
    });
    return compiler;
  }
}
