// @middleware.io/sourcemap-uploader/dist/vite-plugin.js
import {
  __async,
  uploadSourcemaps
} from "./chunk-RPDPQWK5.js";

// src/vite-plugin.ts
import { relative, resolve } from "path";
import { cwd } from "process";
function middlewareVitePlugin(options) {
  const {
    apiKey,
    appVersion,
    basePath = "",
    backendUrl,
    deleteAfterUpload = false
  } = options;
  let resolvedConfig;
  return {
    name: "middleware-sourcemap-upload",
    apply: "build",
    configResolved(config) {
      resolvedConfig = config;
    },
    closeBundle() {
      return __async(this, null, function* () {
        var _a2;
        try {
          const outDir = resolve(resolvedConfig.root, resolvedConfig.build.outDir);
          const path = (_a2 = options.path) != null ? _a2 : relative(cwd(), outDir) || ".";
          yield uploadSourcemaps({
            apiKey,
            appVersion,
            path,
            basePath,
            backendUrl,
            allowNoop: true,
            deleteAfterUpload
          });
        } catch (e) {
          console.error(`Middleware failed to upload sourcemaps: ${e}`);
        }
      });
    }
  };
}
export {
  middlewareVitePlugin as default
};
