// @middleware.io/sourcemap-uploader/dist/webpack-plugin.js
import {
  __async,
  uploadSourcemaps
} from "./chunk-RPDPQWK5.js";

// src/webpack-plugin.ts
var MiddlewareWebpackPlugin = class {
  constructor(apiKey, appVersion, path, basePath, backendUrl, deleteAfterUpload) {
    this.apiKey = apiKey;
    this.appVersion = appVersion;
    this.path = path;
    this.basePath = basePath;
    this.backendUrl = backendUrl;
    this.deleteAfterUpload = deleteAfterUpload;
  }
  apply(compiler) {
    compiler.hooks.afterEmit.tap("MiddlewareWebpackPlugin", () => __async(this, null, function* () {
      try {
        yield uploadSourcemaps({
          apiKey: this.apiKey,
          appVersion: this.appVersion,
          path: this.path,
          basePath: this.basePath,
          backendUrl: this.backendUrl,
          allowNoop: true,
          deleteAfterUpload: this.deleteAfterUpload
        });
      } catch (e) {
        console.error(`Middleware failed to upload sourcemaps: ${e}`);
      }
    }));
    return compiler;
  }
};
export {
  MiddlewareWebpackPlugin as default
};
